node/src/types/block.rs

```diff
@@ -1111,6 +1113,217 @@ impl Item for BlockHeaderWithMetadata {
+                return None;
+            }
+        }
+
+        if !Self::is_continuous_and_descending(&batch, verifiable_chunked_hash_activation) {
+            error!("batch is not continuous");
+            return None;
+        }
+
+        Some(Self(batch))
+    }
+
+    /// Returns inner value.
+    pub fn into_inner(self) -> Vec<BlockHeader> {
+        self.0
+    }
+
+    /// Returns a reference to an inner vector of block headers.
+    pub fn inner(&self) -> &Vec<BlockHeader> {
+        &self.0
+    }
+
+    /// Tests whether the block header batch is continuous and in descending order.
+    pub fn is_continuous_and_descending(
+        batch: &[BlockHeader],
+        verifiable_chunked_hash_activation: EraId,
+    ) -> bool {
+        batch
+            .windows(2)
+            .filter_map(|window| match &window {
+                &[l, r] => Some((l, r)),
+                _ => None,
+            })
+            .all(|(l, r)| {
+                l.height() == r.height() + 1
+                    && l.parent_hash() == &r.hash(verifiable_chunked_hash_activation)
+            })
+    }
+
+    #[cfg(test)]
+    // Test-only constructor allowing creation otherwise invalid data.
+    fn new(batch: Vec<BlockHeader>) -> Self {
+        Self(batch)
+    }
+}
+
+impl Display for BlockHeadersBatch {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "block header batch")
+    }
+}
+
+impl Item for BlockHeadersBatch {
+    type Id = BlockHeadersBatchId;
+
+    type ValidationError = BlockHeadersBatchValidationError;
+
+    const TAG: Tag = Tag::BlockHeaderBatch;
+
+    const ID_IS_COMPLETE_ITEM: bool = false;
+
+    fn validate(
+        &self,
+        verifiable_chunked_hash_activation: EraId,
+    ) -> Result<(), Self::ValidationError> {
+        if self.inner().is_empty() {
+            return Err(BlockHeadersBatchValidationError::BatchEmpty);
+        }
+
+        if !BlockHeadersBatch::is_continuous_and_descending(
+            self.inner(),
+            verifiable_chunked_hash_activation,
+        ) {
+            return Err(BlockHeadersBatchValidationError::BatchNotContinuous);
+        }
+
+        Ok(())
+    }
+
+    fn id(&self, _verifiable_chunked_hash_activation: EraId) -> Self::Id {
+        let upper_batch_height = self.0.first().map(|h| h.height());
+        let lower_batch_height = self.0.last().map(|h| h.height());
+        if lower_batch_height.is_none() || upper_batch_height.is_none() {
+            // ID should be infallible but it is possible that the `Vec` is empty.
+            // In that case we log an error to indicate something went really wrong and use `(0,0)`.
+            error!(
+                ?lower_batch_height,
+                ?upper_batch_height,
+                "received header batch is empty"
+            );
+        }
+        BlockHeadersBatchId::new(
+            upper_batch_height.unwrap_or_default(),
+            lower_batch_height.unwrap_or_default(),
+        )
```

Comment:
I wonder whether the batch's highest block's hash should just be part of the ID? That way, if we're somehow on the wrong chain, the other nodes wouldn't all send us back a whole range of headers that we then throw away due to the mismatch? It would also remove some of the possibly overflowing arithmetic if the ID was just the highest hash and the length.

Reply:
I assumed we would recognize much earlier in the process whether we are on a wrong chain.

Reply:
It depends. If some of our peers are trying to fool us, and we also got the wrong trusted hash?

Reply:
If they try to fool us I don't think this would change much - i.e. they could ignore the request ID and respond with anything they want. 

We do validate the response against the lowest trusted block anyway so we would detect it early on whether the response if valid or not.